.github/actions/nexus-promote-staging/action.yml

```diff
@@ -0,0 +1,83 @@
+name: Promote Nexus staging repository
+description: Close and promote a Nexus Staging Repository
+inputs:
+  version:
+    description: Version being released. Used to update the repository description while closing and releasing
+    required: true
+  staging-repository:
+    description: The id of the staging repository to be promoted
+    required: true
+  nexus-username:
+    description: The Nexus username
+    required: true
+  nexus-password:
+    description: The Nexus password
+    required: true
+  nexus-url:
+    description: The base URL to the Nexus server
+    required: false
+    default: "https://artifacts.alfresco.com/nexus"
+runs:
+  using: composite
+  steps:
+    - name: Close staging repository
+      shell: bash
+      env:
+        NEXUS_USERNAME: ${{ inputs.nexus-username }}
+        NEXUS_PASSWORD: ${{ inputs.nexus-password }}
+        NEXUS_URL: ${{ inputs.nexus-url }}
+        VERSION: ${{ inputs.version }}
+        STAGING_REPOSITORY_ID: ${{ inputs.staging-repository }}
+      run: |
+        CLOSE_STAGING_PAYLOAD_CONTENT=$(envsubst < $GITHUB_ACTION_PATH/close-payload-template.json)
+        curl -u "${NEXUS_USERNAME}":"${NEXUS_PASSWORD}" \
+          -H "Accept: application/json" \
+          -H "Content-Type: application/json" \
+          -d "$CLOSE_STAGING_PAYLOAD_CONTENT" \
+          "${NEXUS_URL}"/service/local/staging/bulk/close
+
+        wait_until_true () {
+            local attempt_counter=0
+            local max_attempts=12
+
+            until "$@"
+            do
+              if [ ${attempt_counter} -eq ${max_attempts} ]
+              then
+                echo "Max attempts reached"
+                exit 1
+              fi
+
+              attempt_counter=$((attempt_counter+1))
+              echo "Condition not reached yet. Attempt $attempt_counter out of $max_attempts. Retrying..."
+              sleep 5
+            done
+        }
+
+        staging_repo_is_closed () {
+          curl -u "${NEXUS_USERNAME}":"${NEXUS_PASSWORD}" \
+            -X GET "${NEXUS_URL}"/service/local/staging/profile_repositories \
+            | yq -p=xml e '.stagingRepositories.data.stagingProfileRepository[]| select (.repositoryId == env(STAGING_REPOSITORY_ID)) | .type ' \
+            | grep open > /dev/null 2>&1
```

Comment:
There is no direct API to get the status of the staging repo?

Reply:
I'm fine with it, I was just wondering.
If it's of any help, I found this in old release notes at https://help.sonatype.com/repomanager2/release-notes/older-release-notes:

```
close/promote/release/drop operations are now asynchronous. Drop operations, including nexus-staging:release with auto-drop enabled, are complete when the response code is 404. For other operations, a GET request to "/service/local/staging/repository/<repo-id>" that returns a 'transitioning" field value of false indicates the operation is complete.
```

Reply:
indeed, this helps: there is an API to retrieve directly the repository using `/service/local/staging/repository/{repositoryId}`. However, in the case of the release I still prefer to use the one returning a list of repositories to avoid handling `404` in the case where the release is completed.